plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JS.xtext

```diff
@@ -581,6 +582,58 @@ PrimaryExpression <Yield> returns Expression:
 	| TemplateLiteral<Yield>
 ;
 
+
+// ****************************************************************************************************
+// see https://facebook.github.io/jsx/
+// ****************************************************************************************************
+
+
+JSXElement:
+	'<' jsxElementName=JSXElementName JSXAttributes
+	(('>' jsxChildren+=JSXChild* JSXClosingElement) | ('/' '>'));
+
+fragment JSXClosingElement *:
+	'<' '/' jsxClosingName=JSXElementName '>';
+
+JSXChild:
+	JSXElement | JSXExpression
+	//	| JSXText -- not supported yet, cf. IDE-2414
+;
+
+// terminal JSX_TEXT: !('{'|'<'|'>'|'}'); needs to be refactored similar to template text, cf. IDE-2414
+
+JSXExpression: '{' expression=AssignmentExpression<false,false> '}';
+
+JSXElementName:
+	expression=JSXElementNameExpression
+	;
+
+JSXElementNameExpression returns Expression:
+	IdentifierRef<false>
+	({ParameterizedPropertyAccessExpression.target=current} ParameterizedPropertyAccessExpressionTail<false>)*
+	//	| JSXNamedspacedName not supported
+;
+
+//JSXNamedspacedName: JSXIdentifier ':' JSXIdentifier -- not supported in N4JSX
+//JSXMemberExpression: JSXIdentifier '.' JSXIdentifier -- defined by means of ParameterizedPropertyAccessExpression
+
+fragment JSXAttributes *:
+	jsxAttributes+=JSXAttribute*;
+
+JSXAttribute:
+	JSXSpreadAttribute
+	|
+	JSXPropertyAttribute;
+
+JSXSpreadAttribute:
+	'{' '...' expression=AssignmentExpression<In=false, Yield=false> '}';
+
+JSXPropertyAttribute:
+	property=[types::IdentifiableElement|IdentifierName] ('=' (jsxAttributeValue=StringLiteral | ('{'
+	jsxAttributeValue=AssignmentExpression<In=false, Yield=false> '}')))?;
```

Comment:
just cosmetic: I'd place the JSX-related grammar at the very bottom of the file, not somewhere in the middle (it is like it's own chapter)

Reply:
done